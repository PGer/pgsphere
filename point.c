#include "point.h"

/*!        
  \file
  This file contains definitions for spherical
  point functions
  \brief definitions of spherical point functions
*/ 


#ifndef DOXYGEN_SHOULD_SKIP_THIS

 PG_FUNCTION_INFO_V1(spherepoint_in);
 PG_FUNCTION_INFO_V1(spherepoint_from_long_lat);
 PG_FUNCTION_INFO_V1(spherepoint_distance);
 PG_FUNCTION_INFO_V1(spherepoint_long);
 PG_FUNCTION_INFO_V1(spherepoint_lat);
 PG_FUNCTION_INFO_V1(spherepoint_x);
 PG_FUNCTION_INFO_V1(spherepoint_y);
 PG_FUNCTION_INFO_V1(spherepoint_z);
 PG_FUNCTION_INFO_V1(spherepoint_xyz);
 PG_FUNCTION_INFO_V1(spherepoint_equal);

#endif

  bool spoint_eq ( const SPoint * p1 , const SPoint * p2 )
  {
    Vector3D a,b ;
    spoint_vector3d ( &a , p1 );
    spoint_vector3d ( &b , p2 );
    return ( vector3d_eq ( &a, &b ) );
  }


  SPoint *  spoint_check (SPoint * spoint)
  {
    static bool lat_is_neg;
    lat_is_neg = ( spoint->lat < 0 )?( TRUE ):( FALSE );

    spoint->lng  = spoint->lng - floor(spoint->lng / ( PID ) )  * PID;
    spoint->lat  = spoint->lat - floor(spoint->lat / ( PID ) )  * PID;
    if( spoint->lng < 0.0 ){
      spoint->lng +=  ( PID );
    }
    if( spoint->lat > PI   ){
      spoint->lat -=  ( 2*PI );
    }
    if( spoint->lat > PIH  ){
      spoint->lat  =  ( PI - spoint->lat );
      spoint->lng +=  ( ( spoint->lng<PI )?( PI ):( -PI ) ) ;
    }
    if( spoint->lat < -PIH ){
      spoint->lat  =  ( -PI - spoint->lat  );
      spoint->lng +=  ( ( spoint->lng<PI )?( PI ):( -PI ) ) ;
    }
    if (
        FPeq(spoint->lat,PIH) && lat_is_neg 
      ) spoint->lat  = -PIH ;
    
    if ( FPeq(spoint->lng,PID) ){
        spoint->lng  = 0.0 ;
    }
    if ( FPzero(spoint->lng) ){
        spoint->lng  = 0.0 ;
    }
    if ( FPzero(spoint->lat) ){
        spoint->lat  = 0.0 ;
    }
    
    return spoint;
  }


  SPoint * vector3d_spoint ( SPoint * p , const Vector3D * v )
  {

    double rho     =  sqrt ( (v->x)*(v->x) + (v->y)*(v->y) );
    if( 0.0 == rho ){
      if( FPzero( v->z ) ) {
        p->lat  = 0.0;
      } else if( v->z > 0 ){
        p->lat  = PIH;
      } else if( v->z < 0 ) {
        p->lat  = - PIH;
      } 
    } else {
      p->lat  = atan( v->z / rho );
    }

    p->lng   = atan2( v->y, v->x );
    if( FPzero(p->lng) ){
    	p->lng = 0.0;
    } else
    if( p->lng < 0.0 ){
    	p->lng += PID ;
    }
    return ( p );
  } 

  Vector3D * spoint_vector3d ( Vector3D * v , const SPoint * p )
  {
    v->x      = cos(p->lng)*cos(p->lat);
    v->y      = sin(p->lng)*cos(p->lat);
    v->z      = sin(p->lat);
    return ( v ) ;
  }

  Datum  spherepoint_in(PG_FUNCTION_ARGS)
  {
    SPoint  * sp  = ( SPoint * ) MALLOC ( sizeof ( SPoint ) ) ;
    char    *  c  = PG_GETARG_CSTRING(0);
    double lng, lat;

    void sphere_yyparse( void );
    init_buffer ( c );
    sphere_yyparse();
    if ( get_point( &lng, &lat ) ){
      sp->lng       = lng;
      sp->lat       = lat;
      spoint_check ( sp );
    } else {
      reset_buffer();
      FREE( sp );
      sp = NULL;
      elog ( ERROR , "spherepoint_in: parse error" );
    }
    reset_buffer();
    PG_RETURN_POINTER( sp );
  }


  Datum  spherepoint_from_long_lat(PG_FUNCTION_ARGS)
  {
    SPoint   * p  = ( SPoint * ) MALLOC ( sizeof ( SPoint ) ) ;
    p->lng        = PG_GETARG_FLOAT8( 0 );
    p->lat        = PG_GETARG_FLOAT8( 1 );
    spoint_check ( p ) ;
    PG_RETURN_POINTER( p );
  }

  float8 spoint_dist ( const SPoint * p1, const SPoint * p2 )
  {
	float8 dl = p1->lng - p2->lng;
    float8 f = (  ( sin( p1->lat )*sin( p2->lat ) + cos( p1->lat )*cos( p2->lat )*cos( dl ) ) );
    if( FPeq( f, 1.0 ) ){
   	  /* for small distances */
      Vector3D v1, v2, v3;
      spoint_vector3d(&v1, p1);
      spoint_vector3d(&v2, p2);
      vector3d_cross( &v3, &v1, &v2 );
      f = vector3d_length(&v3);
    } else {
      f = acos(f);
    }
    if ( FPzero(f) ){
      return 0.0;
    } else {
      return f;
    }
  }

  Datum spherepoint_distance(PG_FUNCTION_ARGS) 
  {

    SPoint  * p1 =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    SPoint  * p2 =  ( SPoint * ) PG_GETARG_POINTER ( 1 ) ;
    PG_RETURN_FLOAT8 ( spoint_dist ( p1, p2 ) );

  }

  Datum  spherepoint_long(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * )  PG_GETARG_POINTER ( 0 ) ;
    PG_RETURN_FLOAT8 ( p->lng ); 
  }

  Datum  spherepoint_lat(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    PG_RETURN_FLOAT8 ( p->lat ); 
  }

  Datum  spherepoint_x(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.x   ); 
  }
  
  Datum  spherepoint_y(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.y   ); 
  }

  Datum  spherepoint_z(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.z   ); 
  }

  Datum  spherepoint_xyz(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    Datum    dret[3];
    ArrayType  *result;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    dret[0] = Float8GetDatumFast(v.x);
    dret[1] = Float8GetDatumFast(v.y);
    dret[2] = Float8GetDatumFast(v.z);
	#ifdef FLOAT8PASSBYVAL
		result = construct_array ( dret , 3, FLOAT8OID, sizeof(float8), FLOAT8PASSBYVAL, 'd');
	#else
        result = construct_array ( dret , 3, FLOAT8OID, sizeof(float8), false /* float8 byval */ , 'd' );
	#endif
    PG_RETURN_ARRAYTYPE_P(result);                                                                                                          
  }

  Datum  spherepoint_equal(PG_FUNCTION_ARGS)
  {
    SPoint  * p1 =  ( SPoint * )  PG_GETARG_POINTER ( 0 ) ;
    SPoint  * p2 =  ( SPoint * )  PG_GETARG_POINTER ( 1 ) ;
    PG_RETURN_BOOL( spoint_eq ( p1 , p2 ) );
  }

