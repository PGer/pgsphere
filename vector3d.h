#ifndef __PGS_VECTOR3D_H__
#define __PGS_VECTOR3D_H__

#include "pg_sphere.h"

/*!
  \file
  \brief Vector declarations
*/


/*!
  The definition of a three dimensional
  vector.
  \brief three-dimensional vector
*/
 typedef struct {
    float8    x;         //!< x value ( -1.0 .. 1.0 )
    float8    y;         //!< y value ( -1.0 .. 1.0 )
    float8    z;         //!< z value ( -1.0 .. 1.0 ) 
 } Vector3D;


/*!
  \brief calculate the cross product of two vectors
  \param v1  pointer to first vector
  \param v2  pointer to second vector
  \param out pointer to result vector
  \return pointer to result vector
*/
Vector3D * vector3d_cross     ( Vector3D * out , const Vector3D * v1 , const Vector3D * v2 );

/*!
  \brief checks equalness of two vectors
  \param a pointer to first vector
  \param b pointer to second vector
  \return true, if equal
*/
bool       vector3d_eq        ( const Vector3D * a   , const Vector3D * b  );


/*!
  \brief calculate the scalar product of two vectors
  \param v1  pointer to first vector
  \param v2  pointer to second vector
  \param out pointer to scalar product
  \return pointer to scalar product
*/
float8  * vector3d_scalar ( float8 * out , Vector3D * v1 , Vector3D * v2 );

#endif
