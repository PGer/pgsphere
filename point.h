#ifndef __PGS_POINT_H__
#define __PGS_POINT_H__

#include "vector3d.h"
#include "sbuffer.h"
/*!
  \file
  This file contains declaration for spherical
  point and functions
  \brief spherical point declarations
*/

/*!
  The definition of spherical point.
  \brief Spherical point
*/
  typedef struct {
   float8 lng;  //!< longitude value in radians
   float8 lat;  //!< latitude value in radians
  }  SPoint;

  /*!
    Calculate the distance of two spherical points
    \brief Distance of two spherical points
    \return distance in radians
  */
  float8 spoint_dist ( const SPoint * p1, const SPoint * p2 );


  /*!
    \brief check whether two points are equal
    \param p1 first point
    \param p2 second point
    \return true, if equal
  */
  bool spoint_eq ( const SPoint * p1 , const SPoint * p2 );

  /*!
    Checks the longitude and latitude values and take sure
    the right value ranges
    \brief  validate the spherical point
    \param  spoint pointer to spherical point
    \return pointer to spherical point
  */
  SPoint *  spoint_check (SPoint * spoint);

  /*!
    \brief transforms a 3 dim.vector to a spherical point 
    \param p pointer to spherical point
    \param v pointer to 3 dim. vector
    \return pointer to spherical point
  */
  SPoint   * vector3d_spoint    ( SPoint   * p   , const Vector3D * v );

  /*!
    \brief transforms a spherical point to a 3 dim.vector
    \param v pointer to 3 dim. vector
    \param p pointer to spherical point
    \return pointer to 3 dim. vector
  */
  Vector3D * spoint_vector3d    ( Vector3D * v   , const SPoint   * p );


  /*!
    Take the input and stores it as a spherical point
    \return a spherical point datum
    \note Does check the input too.
   */
  Datum  spherepoint_in(PG_FUNCTION_ARGS);


  /*!
    Create a spherical point from longitude
    and latitude both in radians
    \brief  point created from longitude and latitude
    \return a spherical point datum
    \note PostgreSQL function
   */
  Datum  spherepoint_from_long_lat(PG_FUNCTION_ARGS);


  /*!
    Calculate the distance of two spherical points
    \brief Distance of two spherical points
    \return distance ( float8 ) as datum
    \see spoint_dist ( SPoint *, SPoint * )
    \note PostgreSQL function
  */
  Datum spherepoint_distance(PG_FUNCTION_ARGS);

  /*!
    \brief longitude of spherical point
    \return longitude datum in radians ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_long(PG_FUNCTION_ARGS);

  /*!
    \brief latitude of spherical point
    \return latitude datum in radians ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_lat(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian x-value of spherical point
    \return x-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_x(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian y-value of spherical point
    \return y-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_y(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian z-value of spherical point
    \return z-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_z(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian values of spherical point
    \return array datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_xyz(PG_FUNCTION_ARGS);

  /*!
    \brief check whether two points are equal
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_equal(PG_FUNCTION_ARGS);


#endif
