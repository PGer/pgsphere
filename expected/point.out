-- radian input
SELECT spoint (-3*pi(),       0);
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint ( 3*pi(),       0);
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint (      0,    pi());
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint (      0,   -pi());
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint ( 2*pi(),  2*pi());
 spoint  
---------
 (0 , 0)
(1 row)

SELECT spoint (   pi(),  2*pi());
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint (      0,  pi()/2);
        spoint         
-----------------------
 (0 , 1.5707963267949)
(1 row)

SELECT spoint (      0, -pi()/2);
         spoint         
------------------------
 (0 , -1.5707963267949)
(1 row)

-- degree input
SELECT spoint '(-540d,    0d)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '( 540d,    0d)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(   0d,  180d)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(   0d, -180d)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '( 360d,  360d)';
 spoint  
---------
 (0 , 0)
(1 row)

SELECT spoint '( 180d,  360d)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(   0d,   90d)';
        spoint         
-----------------------
 (0 , 1.5707963267949)
(1 row)

SELECT spoint '(   0d,  -90d)';
         spoint         
------------------------
 (0 , -1.5707963267949)
(1 row)

-- dms input
SELECT spoint '( 0.0d   0m  0s,    0d 0m 0s)';
ERROR:  parsing: syntax error
SELECT spoint '(   0d  60m  0s,    0d 0m 0s)';
ERROR:  parsing: minutes or seconds invalid (min:60.000000 sec:0.000000)
SELECT spoint '(   0d   0m 60s,    0d 0m 0s)';
ERROR:  parsing: minutes or seconds invalid (min:0.000000 sec:60.000000)
SELECT spoint '( 539d 59m 59.9999s,    0d 0m 0s)';
         spoint         
------------------------
 (3.14159265310498 , 0)
(1 row)

SELECT spoint '(   0d  0m  0s,       180d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(   0d  0m  0s,      -180d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '( 360d  0m  0s,       360d 0m 0s)';
 spoint  
---------
 (0 , 0)
(1 row)

SELECT spoint '( 180d  0m  0s,       360d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(   0d  0m  0s,        90d 0m 0s)';
        spoint         
-----------------------
 (0 , 1.5707963267949)
(1 row)

SELECT spoint '(   0d  0m  0s,       -90d 0m 0s)';
         spoint         
------------------------
 (0 , -1.5707963267949)
(1 row)

-- hms input for latitude
SELECT spoint '( 0.0h   0m  0s,    0)';
ERROR:  parsing: syntax error
SELECT spoint '(   0h  60m  0s,    0)';
ERROR:  parsing: minutes or seconds invalid (min:60.000000 sec:0.000000)
SELECT spoint '(   0h   0m 60s,    0)';
ERROR:  parsing: minutes or seconds invalid (min:0.000000 sec:60.000000)
SELECT spoint '( 35h 59m 59.9999s,    0d 0m 0s)';
         spoint         
------------------------
 (3.14159264631759 , 0)
(1 row)

SELECT spoint '(  0h  0m  0s,       180d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(  0h  0m  0s,      -180d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '( 24h  0m  0s,       360d 0m 0s)';
 spoint  
---------
 (0 , 0)
(1 row)

SELECT spoint '( 12h  0m  0s,       360d 0m 0s)';
         spoint         
------------------------
 (3.14159265358979 , 0)
(1 row)

SELECT spoint '(  0h  0m  0s,        90d 0m 0s)';
        spoint         
-----------------------
 (0 , 1.5707963267949)
(1 row)

SELECT spoint '(  0h  0m  0s,       -90d 0m 0s)';
         spoint         
------------------------
 (0 , -1.5707963267949)
(1 row)

-- mixed input
-- operators
SELECT spoint '(0, 90d)'   = spoint '(0, 90d)';
 ?column? 
----------
 t
(1 row)

SELECT spoint '(0, 90d)'   = spoint '(0,-90d)';
 ?column? 
----------
 f
(1 row)

SELECT spoint '(0,-90d)'   = spoint '(0,-90d)';
 ?column? 
----------
 t
(1 row)

SELECT spoint '(0, 90d)'  != spoint '(0, 90d)';
 ?column? 
----------
 f
(1 row)

SELECT spoint '(0, 90d)'  != spoint '(0,-90d)';
 ?column? 
----------
 t
(1 row)

SELECT spoint '(0,-90d)'  != spoint '(0,-90d)';
 ?column? 
----------
 f
(1 row)

SELECT spoint '(0d,  0)'   = spoint '(360d,0)';
 ?column? 
----------
 t
(1 row)

